R/resample_draws.R

```diff
@@ -0,0 +1,176 @@
+#' Possible choices are `"simple"` for simple random resampling with
+#' replacement, `"simple_no_replace"` for simple random resampling without
+#' replacement, `"stratified"` for stratified resampling with replacement
+#' and `"deterministic"` for deterministic resampling with replacement.
+#' Currently, `"stratified"` is the default as it has comparably
+#' low variance and bias with respect to ideal resampling. The latter would
+#' sample perfectly proportional to the weights but is not possible in practice
+#' where only a finite number of draws are available. For more details
+#' about resampling methods, see Kitagawa (1996).
+#' @param ndraws Number of draws to be returned. By default (`NULL`),
+#' `ndraws` is set internally to the total number of draws in `x` if sensible.
+#' @template args-methods-dots
+#' @template return-draws
+#'
+#' @details It is only possible to resample `draws` formats in which draws
+#' can be subsetted directly without breaking the format's structure,
+#' that is, only in `draws_matrix` and `draws_df` objects. Upon usage of
+#' `resample_draws`, chain information will be silently dropped due to
+#' subsetting of individual draws (see [`subset_draws`] for details).
+#'
+#' @references
+#' Kitagawa, G., Monte Carlo Filter and Smoother for Non-Gaussian Nonlinear '
+#' State Space Models, *Journal of Computational and Graphical Statistics*,
+#' 5(1):1-25, 1996.
+#'
+#' @examples
+#' x <- as_draws_df(example_draws())
+#'
+#' # random weights for illustrationary purposes
+#' w <- runif(ndraws(x), 0, 10)
+#'
+#' # use default stratified sampling
+#' x_rs <- resample_draws(x, weights = w)
+#' summarise_draws(x_rs, default_summary_measures())
+#'
+#' # use simple random sampling
+#' x_rs <- resample_draws(x, weights = w, method = "simple")
+#' summarise_draws(x_rs, default_summary_measures())
+#'
+#' @export
+resample_draws <- function(x, ...) {
+  UseMethod("resample_draws")
+}
+
+#' @rdname resample_draws
+#' @export
+resample_draws.draws <- function(x, weights, method = "stratified",
+                                 ndraws = NULL, ...) {
+  if (!allow_resample_draws(x)) {
+    stop2("'resample_draws' requires formats which can be subsetted ",
+          "by individual draws, not only by iterations or chains.")
+  }
+  ndraws_total <- ndraws(x)
+  assert_numeric(weights, len = ndraws_total, lower = 0, null.ok = TRUE)
+  assert_choice(method, supported_resample_methods())
+  assert_number(ndraws, null.ok = TRUE, lower = 0)
+  weights <- weights / sum(weights)
+  method_fun <- paste0(".resample_", method)
+  method_fun <- get(method_fun, asNamespace("posterior"))
+  sel_draws <- method_fun(
+    weights = weights,
+    ndraws = ndraws,
+    ...
+  )
+  subset_draws(x, draw = sel_draws, unique = FALSE)
+}
+
+# simple random resampling with replacement
+# @return index vector of length 'ndraws'
+.resample_simple <- function(weights, ndraws, ...) {
+  ndraws <- default_ndraws_resample(ndraws, weights)
```

Comment:
Done.

Reply:
The reason I was initially handling this different was because `ndraws` default cannot be set for method `simple_no_replace`. I think the new way of handling this exception is now cleaner.